Assign ServerConnection to Player before processing post-login packets

wallenjos01 opened the pull request:
Currently, when transitioning to a new server, the proxy begins auto-reading (and therefore processing) packets from the pending server connection before assigning that connection to the player via `ConnectedPlayer::setConnectedServer`. This causes a race condition if the backend server sends certain packets (particularly plugin messages) right after sending a `JoinGamePacket`. For example, the following has inconsistent behavior depending on when a plugin message is sent:
```java
@Subscribe
public void onPluginMessage(PluginMessageEvent event) {
    if(event.getSource() instanceof ServerConnection sc) {

        // Will work consistently
        sc.sendPluginMessage(MinecraftChannelIdentifier.create("test", "message"), "Hello, World".getBytes()); 
        
        // Will often fail if the backend server sends a plugin message immediately after join.
        sc.getPlayer()
            .getCurrentServer()
            .orElseThrow()
            .sendPluginMessage(MinecraftChannelIdentifier.create("test", "message"), "Hello, World".getBytes());
    }
}
```
This PR reverses the order of two lines in `TransitionSessionHandler`, making the proxy wait to process packets sent by a backend server until after assigning the player's connection.

wallenjos01 committed:
Assign server connection before processing post-login packets

kashike commented on proxy/src/main/java/com/velocitypowered/proxy/connection/backend/TransitionSessionHandler.java:
```diff
@@ -137,12 +137,12 @@ public boolean handle(JoinGamePacket packet) {
           serverConn.getPlayer().setConnectedServer(serverConn);
 
+          // Clean up disabling auto-read while the connected event was being processed.
+          smc.setAutoReading(true);
```
Would you mind expanding the comment above this line to include your description (or similar)?

Gerrygames commented:
Pretty sure that this does not make a difference. The code in question runs on the event loop of the server connection, so no packets from that connection can be handled during its execution.

electronicboy commented:
Has this been tested? Because I'd imagine that the more bigger issue here is going to be that disabling auto-read doesn't promise to disable reading *everything*, and so there can still be stuff which ends up being processed, at least that is my understanding; The auto-reading re-enable + field update both occur within the same area.

wallenjos01 committed:
Document reasoning for waiting to auto-read

wallenjos01 commented:
> Has this been tested? 

I created this PR because I ran into errors caused by this bug while working on a plugin for my private server, but just to be sure, I spent some time writing a plugin to test it. I tested with both Velocity build 385, downloaded from the website, and my own fork. I tested both Spigot 1.20.5 and Fabric 1.20.5 backends. I got about a 40% error rate when using build 385, and a 0% error rate using my fork. I have attached the relevant logs and code.

[pmtest.zip](https://github.com/PaperMC/Velocity/files/15142692/pmtest.zip)

Pantera07 committed:
Add version information for 1.20.6 (#1310)

https://www.minecraft.net/en-us/article/minecraft-java-edition-1-20-6

electronicboy committed:
Properly pass connection exceptions to the logger

electronicboy committed:
Do not send empty upsert player info packets

There is no harm in this, but seemingly, some common mod blows up on this, and,
well, why eat buffer?

electronicboy committed:
Hotfix cancelling unsigned 1.20.6 commands

R00tB33rMan commented on proxy/src/main/java/com/velocitypowered/proxy/connection/backend/TransitionSessionHandler.java:
```diff
@@ -137,12 +137,14 @@ public boolean handle(JoinGamePacket packet) {
           smc.setActiveSessionHandler(StateRegistry.PLAY,
               new BackendPlaySessionHandler(server, serverConn));
 
+          // Now set the connected server. 
```
Trim whitespace after the period (line 140)

wallenjos01 committed:
Cleanup comment

wallenjos01 committed:
Merge branch 'PaperMC:dev/3.0.0' into dev/3.0.0

kennytv reviewed (approved):
From a quick look, this seems to be caused by the manual channel.read call in MinecraftConnection#setAutoReading, so the PR should be good
